### ISSUE TITLE
Html Template Resource not working on Dynamics Apps

### ISSUE THREAD

**Issue by alexmare93** [DigitalFlow/Xrm-Html-Templating#16]
Hi,

I've noticed that when creating an App on Dynamics 9.0 On-Prem and inserting HTML Template Entity in it, when trying to open an HTML Template record the HTML editor does not appear.
Watching in F12, this is the error:

```
CrashReport:

		FileName:/%7B637245272150000009%7D/WebResources/oss_/XrmHtmlTemplating/js/xrm_html_templating.js

		LineNumber:8251

		Function:anonymousror:JSON.parse:unexpectedcharacteratline1column1oftheJSONdata

		ErrorReport:<CrmScriptErrorReport>

  <ReportVersion>1.0</ReportVersion>

  <ScriptErrorDetails>

   <Message>SyntaxError&#58; JSON.parse&#58; unexpected character at line 1 column 1 of the JSON data</Message>

   <Line>8251</Line>

   <URL>&#47;&#37;7B637245272150000009&#37;7D&#47;WebResources&#47;oss_&#47;XrmHtmlTemplating&#47;js&#47;xrm_html_templating.js&#63;ver&#61;0</URL>

   <PageURL>&#47;&#37;7B637245272150000009&#37;7D&#47;WebResources&#47;oss_&#47;XrmHtmlTemplating&#47;html&#47;xrm_html_templating.html&#63;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;data&#61;&#37;7b&#37;22jsonField&#37;22&#37;3a&#37;20&#37;22oss_json&#37;22&#37;2c&#37;20&#37;22htmlField&#37;22&#37;3a&#37;20&#37;22oss_html&#37;22&#37;7d</PageURL>

   <Function>anonymousror&#58;JSON.parse&#58;unexpectedcharacteratline1column1oftheJSONdata</Function>

   <FunctionRaw>SyntaxError&#58; JSON.parse&#58; unexpected character at line 1 column 1 of the JSON data</FunctionRaw>

   <CallStack>

    <Function>anonymousror&#58;JSON.parse&#58;unexpectedcharacteratline1column1oftheJSONdata</Function>

   </CallStack>

  </ScriptErrorDetails>

  <ClientInformation>

   <BrowserUserAgent>Mozilla&#47;5.0 &#40;Windows NT 10.0&#59; Win64&#59; x64&#59; rv&#58;75.0&#41; Gecko&#47;20100101 Firefox&#47;75.0</BrowserUserAgent>

   <BrowserLanguage>undefined</BrowserLanguage>

   <SystemLanguage>undefined</SystemLanguage>

   <UserLanguage>undefined</UserLanguage>

   <ScreenResolution>1920x1080</ScreenResolution>

   <ClientName>Web</ClientName>

   <ClienState>Online</ClienState>

   <ClientTime>2020-05-08T12&#58;36&#58;17</ClientTime>

  </ClientInformation>

</CrmScriptErrorReport>
```
How can I solve it?

Thank you

**Comment by DigitalFlow** [comment]
Hi @alexmare93,

can you please check the content of the "oss_json" field on your form when the error appears?
It should either contain nothing or a valid json string which will be parsed.

In your case there seems to be some invalid value inside that field on load.

Kind Regards,
Florian

**Comment by alexmare93** [comment]
Hi Florian,

This is the content of the json:

{"counters":{"u_column":6,"u_row":3,"u_content_image":1,"u_content_text":2},"body":{"rows":[{"cells":[1],"columns":[{"contents":[{"type":"image","values":{"containerPadding":"0px","_meta":{"htmlID":"u_content_image_1","htmlClassNames":"u_content_image"},"selectable":true,"draggable":true,"deletable":true,"src":{"url":"http://img.comoliferrari.it/PORTALE%20CF/COMOLI%20FERRARI.png","width":755,"height":195},"textAlign":"center","altText":"Image","action":{"name":"web","values":{"href":"","target":"_blank"},"url":"","target":"_self"},"fullWidth":false,"maxWidth":"100%"}}],"values":{"backgroundColor":"","padding":"0px","border":{},"_meta":{"htmlID":"u_column_2","htmlClassNames":"u_column"}}}],"values":{"columns":false,"backgroundColor":"#008cca","columnsBackgroundColor":"","backgroundImage":{"url":"","fullWidth":true,"repeat":false,"center":true,"cover":false},"padding":"0px","hideDesktop":false,"hideMobile":false,"noStackMobile":false,"_meta":{"htmlID":"u_row_2","htmlClassNames":"u_row"},"selectable":true,"draggable":true,"deletable":true}},{"cells":[1],"columns":[{"contents":[{"type":"text","values":{"containerPadding":"10px","_meta":{"htmlID":"u_content_text_1","htmlClassNames":"u_content_text"},"selectable":true,"draggable":true,"deletable":true,"color":"#000","textAlign":"left","lineHeight":"140%","linkStyle":{"inherit":true,"linkColor":"#0000ee","linkHoverColor":"#0000ee","linkUnderline":true,"linkHoverUnderline":true},"text":"<p style=\"font-size: 14px; line-height: 140%; text-align: left;\"><span style=\"font-size: 14px; line-height: 19.6px; color: #000000;\"><span style=\"line-height: 19.6px; font-family: helvetica, sans-serif; font-size: 14px;\">Gentile&nbsp;<span style=\"color: #24292e; font-size: 14px; line-height: 19.6px;\">${{Value(\"regardingobjectid.std_contactid\")}}</span>,</span></span></p>\n<p style=\"font-size: 14px; line-height: 140%; text-align: left;\">&nbsp;</p>\n<p style=\"font-size: 14px; line-height: 140%; text-align: left;\"><span style=\"font-family: helvetica, sans-serif; font-size: 14px; line-height: 19.6px;\">Grazie per esserti registrato all'evento <strong>${{Value(\"regardingobjectid.std_sessionid\")}}</strong>.</span></p>\n<p style=\"font-size: 14px; line-height: 140%; text-align: left;\">&nbsp;</p>\n<p style=\"font-size: 14px; line-height: 140%; text-align: left;\">Cordiali saluti,</p>\n<p style=\"font-size: 14px; line-height: 140%; text-align: left;\">Comoli, Ferrari &amp; C. spa</p>"}}],"values":{"_meta":{"htmlID":"u_column_1","htmlClassNames":"u_column"}}}],"values":{"columns":false,"backgroundColor":"#ffffff","columnsBackgroundColor":"","backgroundImage":{"url":"","fullWidth":true,"repeat":false,"center":true,"cover":false},"padding":"10px","hideDesktop":false,"hideMobile":false,"noStackMobile":false,"_meta":{"htmlID":"u_row_1","htmlClassNames":"u_row"},"selectable":true,"draggable":true,"deletable":true}},{"cells":[1],"columns":[{"contents":[{"type":"text","values":{"containerPadding":"10px","_meta":{"htmlID":"u_content_text_2","htmlClassNames":"u_content_text"},"selectable":true,"draggable":true,"deletable":true,"color":"#000","textAlign":"left","lineHeight":"140%","linkStyle":{"inherit":true,"linkColor":"#0000ee","linkHoverColor":"#0000ee","linkUnderline":true,"linkHoverUnderline":true},"text":"<p style=\"font-size: 14px; line-height: 140%; text-align: left;\"><a href=\"https://www.comoliferrari.it/\"><strong>Comoli Ferrari &amp; c. spa</strong></a> <br /><span style=\"font-size: 10px; line-height: 14px;\">Sede Legale e Direzione: Via Enrico Mattei 4 - 28100 NOVARA</span><br /><span style=\"font-size: 10px; line-height: 14px;\"> Cap.Soc. &euro; 5.100.000,00 int.vers. | R.I. di Novara n&deg; 00123060030 | R.E.A. n&deg; 33928&nbsp; &nbsp; &nbsp;&nbsp;</span></p>"}}],"values":{"backgroundColor":"","padding":"0px","border":{},"_meta":{"htmlID":"u_column_6","htmlClassNames":"u_column"}}}],"values":{"columns":false,"backgroundColor":"#ffdd00","columnsBackgroundColor":"","backgroundImage":{"url":"","fullWidth":true,"repeat":false,"center":true,"cover":false},"padding":"0px","hideDesktop":false,"hideMobile":false,"noStackMobile":false,"_meta":{"htmlID":"u_row_3","htmlClassNames":"u_row"},"selectable":true,"draggable":true,"deletable":true}}],"values":{"backgroundColor":"#e7e7e7","backgroundImage":{"url":"","fullWidth":true,"repeat":false,"center":true,"cover":false},"contentWidth":"500px","fontFamily":{"label":"Arial","value":"arial,helvetica,sans-serif"},"linkStyle":{"body":true,"linkColor":"#0000ee","linkHoverColor":"#0000ee","linkUnderline":true,"linkHoverUnderline":true},"_meta":{"htmlID":"u_body","htmlClassNames":"u_body"}}}}

From the default sitemap, this record opens with the HTML editor; from the dedicated app it doesn’t.

By viewing the object explorer, I get these errors:

CrashReport:
                              FileName:/form/ClientApiWrapper.aspx
                              LineNumber:1
                              Function:anonymousror:Unexpectedtoken'<'
                              ErrorReport:<CrmScriptErrorReport>
  <ReportVersion>1.0</ReportVersion>
  <ScriptErrorDetails>
   <Message>Uncaught SyntaxError&#58; Unexpected token &#39;&#60;&#39;</Message>
   <Line>1</Line>
   <URL>&#47;form&#47;ClientApiWrapper.aspx&#63;ver&#61;1456360070</URL>
   <PageURL>&#47;form&#47;page.aspx&#63;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;lcid&#61;1033&#38;themeId&#61;5f4ca6a8-cba2-46f0-ca12-69183bac8187&#38;tstamp&#61;846105562&#38;updateTimeStamp&#61;636802264284388832&#38;userts&#61;132336378868349104&#38;ver&#61;1456360070&#35;etc&#61;10566&#38;pagetype&#61;entityrecord&#38;extraqs&#61;&#37;3fetc&#37;3d10566&#37;26id&#37;3d&#37;257b0EC255FF-EE57-EA11-83AE-005056AF1EE7&#37;257d&#37;26rskey&#37;3d&#37;257b57467703-4C97-4D90-A86B-B28DAC38DD15&#37;257d&#37;26_gridType&#37;3d10566&#38;rskey&#61;&#37;7b57467703-4C97-4D90-A86B-B28DAC38DD15&#37;7d&#38;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;pagemode&#61;iframe&#38;counter&#61;1589181082101</PageURL>
   <Function>anonymousror&#58;Unexpectedtoken&#39;&#60;&#39;</Function>
   <FunctionRaw>SyntaxError&#58; Unexpected token &#39;&#60;&#39;</FunctionRaw>
   <CallStack>
    <Function>anonymousror&#58;Unexpectedtoken&#39;&#60;&#39;</Function>
   </CallStack>
  </ScriptErrorDetails>
  <ClientInformation>
   <BrowserUserAgent>Mozilla&#47;5.0 &#40;Windows NT 10.0&#59; Win64&#59; x64&#41; AppleWebKit&#47;537.36 &#40;KHTML, like Gecko&#41; Chrome&#47;81.0.4044.138 Safari&#47;537.36</BrowserUserAgent>
   <BrowserLanguage>undefined</BrowserLanguage>
   <SystemLanguage>undefined</SystemLanguage>
   <UserLanguage>undefined</UserLanguage>
   <ScreenResolution>1920x1080</ScreenResolution>
   <ClientName>Web</ClientName>
   <ClienState>Online</ClienState>
   <ClientTime>2020-05-11T09&#58;11&#58;25</ClientTime>
  </ClientInformation>
</CrmScriptErrorReport>


CrashReport:
                              FileName:
                              LineNumber:0
                              Function:
                              ErrorReport:<CrmScriptErrorReport>
  <ReportVersion>1.0</ReportVersion>
  <ScriptErrorDetails>
   <Message>Script error.</Message>
   <Line>0</Line>
   <URL></URL>
   <PageURL>&#47;&#37;7B637245272150000009&#37;7D&#47;WebResources&#47;oss_&#47;XrmHtmlTemplating&#47;html&#47;xrm_html_templating.html&#63;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;data&#61;&#37;7b&#37;22jsonField&#37;22&#37;3a&#37;20&#37;22oss_json&#37;22&#37;2c&#37;20&#37;22htmlField&#37;22&#37;3a&#37;20&#37;22oss_html&#37;22&#37;7d</PageURL>
   <Function></Function>
   <FunctionRaw></FunctionRaw>
   <CallStack>
   </CallStack>
  </ScriptErrorDetails>
  <ClientInformation>
   <BrowserUserAgent>Mozilla&#47;5.0 &#40;Windows NT 10.0&#59; Win64&#59; x64&#41; AppleWebKit&#47;537.36 &#40;KHTML, like Gecko&#41; Chrome&#47;81.0.4044.138 Safari&#47;537.36</BrowserUserAgent>
   <BrowserLanguage>undefined</BrowserLanguage>
   <SystemLanguage>undefined</SystemLanguage>
   <UserLanguage>undefined</UserLanguage>
   <ScreenResolution>1920x1080</ScreenResolution>
   <ClientName>Web</ClientName>
   <ClienState>Online</ClienState>
   <ClientTime>2020-05-11T09&#58;11&#58;26</ClientTime>
  </ClientInformation>
</CrmScriptErrorReport>

Thank you,

ALESSANDRO MARENGHI   |  Dynamics 365 (CE+CS+FS) Developer

T +39 347 5937193 | alessandro.marenghi@porini.it<mailto:alessandro.marenghi@porini.it>

[logo-dgs-porini-1]<https://eur02.safelinks.protection.outlook.com/?url=http%3A%2F%2Fwww.porini.it%2F&data=02%7C01%7Calessandro.marenghi%40porini.it%7Cbe054de5d8374572116508d7e5c8f78a%7Cb3bce17afac9410ab1b532bbd3e015f5%7C0%7C0%7C637230523622681409&sdata=t13GwoW9AK21aqseiQm%2BMnarEyXbfXKZm7eqteRmG%2B0%3D&reserved=0>

CONFIDENTIALITY NOTICE: This message and its attachments (if any) may contain confidential, proprietary or legally privileged information and is intended only for the use of the addressee named above. No confidentiality or privilege is waived or lost by any mistransmission. If you are not the intended recipient of this message you are hereby notified that you must not use, disseminate, copy it in any form or take any action in reliance on it. If you have received this message in error please delete it and any copies of it and kindly inform the sender of this e-mail.

**Comment by DigitalFlow** [comment]
Hi @alexmare93,

So did you create a custom form or are you using the default one?
Did you modify the IFrame inside the form?

I have quite a few customers who use this module in UCI and have not reported such errors until now.
It almost seems as if the solution was trying to parse the HTML data as json, which could occur if you mixed up the IFrame parameters.

Kind Regards,
Florian

**Comment by alexmare93** [comment]
Hi Florian,

We’re using the default form, without touching anything that could be custom in that (Iframe, fields, sections etc).
The matter is that we don’t use the UCI, but we use an App (like it was in version 8.2) in version 9.0 On-Prem, which maintained the same interface of earlier version. I think that it is something in it. I show you some pics:

This one is from the default sitemap:

[cid:image002.jpg@01D6291D.E653A480]


By going here
[cid:image012.jpg@01D6291D.E653A480]

Then here:
[cid:image013.jpg@01D6291D.E653A480]

Something changes:
[cid:image014.jpg@01D6291D.E653A480]

The error is always this one:

CrashReport:
                              FileName:/form/ClientApiWrapper.aspx
                              LineNumber:1
                              Function:anonymousror:Unexpectedtoken'<'
                              ErrorReport:<CrmScriptErrorReport>
  <ReportVersion>1.0</ReportVersion>
  <ScriptErrorDetails>
   <Message>Uncaught SyntaxError&#58; Unexpected token &#39;&#60;&#39;</Message>
   <Line>1</Line>
   <URL>&#47;form&#47;ClientApiWrapper.aspx&#63;ver&#61;1456360070</URL>
   <PageURL>&#47;form&#47;page.aspx&#63;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;lcid&#61;1033&#38;themeId&#61;5f4ca6a8-cba2-46f0-ca12-69183bac8187&#38;tstamp&#61;856167998&#38;updateTimeStamp&#61;636802264284388832&#38;userts&#61;132338358600820831&#38;ver&#61;1456360070&#35;etc&#61;10566&#38;pagetype&#61;entityrecord&#38;extraqs&#61;&#37;3fetc&#37;3d10566&#37;26id&#37;3d&#37;257b0EC255FF-EE57-EA11-83AE-005056AF1EE7&#37;257d&#37;26rskey&#37;3d&#37;257b57467703-4C97-4D90-A86B-B28DAC38DD15&#37;257d&#37;26_gridType&#37;3d10566&#38;rskey&#61;&#37;7b57467703-4C97-4D90-A86B-B28DAC38DD15&#37;7d&#38;appid&#61;a369bcba-57a2-409f-8b8b-9f61bb878586&#38;pagemode&#61;iframe&#38;counter&#61;1589363761905</PageURL>
   <Function>anonymousror&#58;Unexpectedtoken&#39;&#60;&#39;</Function>
   <FunctionRaw>SyntaxError&#58; Unexpected token &#39;&#60;&#39;</FunctionRaw>
   <CallStack>
    <Function>anonymousror&#58;Unexpectedtoken&#39;&#60;&#39;</Function>
   </CallStack>
  </ScriptErrorDetails>
  <ClientInformation>
   <BrowserUserAgent>Mozilla&#47;5.0 &#40;Windows NT 10.0&#59; Win64&#59; x64&#41; AppleWebKit&#47;537.36 &#40;KHTML, like Gecko&#41; Chrome&#47;81.0.4044.138 Safari&#47;537.36</BrowserUserAgent>
   <BrowserLanguage>undefined</BrowserLanguage>
   <SystemLanguage>undefined</SystemLanguage>
   <UserLanguage>undefined</UserLanguage>
   <ScreenResolution>1920x1080</ScreenResolution>
   <ClientName>Web</ClientName>
   <ClienState>Online</ClienState>
   <ClientTime>2020-05-13T11&#58;56&#58;02</ClientTime>
  </ClientInformation>
</CrmScriptErrorReport>


Thank you for your support,

ALESSANDRO MARENGHI   |  Dynamics 365 (CE+CS+FS) Developer

T +39 347 5937193 | alessandro.marenghi@porini.it<mailto:alessandro.marenghi@porini.it>

[logo-dgs-porini-1]<https://eur02.safelinks.protection.outlook.com/?url=http%3A%2F%2Fwww.porini.it%2F&data=02%7C01%7Calessandro.marenghi%40porini.it%7Cbe054de5d8374572116508d7e5c8f78a%7Cb3bce17afac9410ab1b532bbd3e015f5%7C0%7C0%7C637230523622681409&sdata=t13GwoW9AK21aqseiQm%2BMnarEyXbfXKZm7eqteRmG%2B0%3D&reserved=0>

CONFIDENTIALITY NOTICE: This message and its attachments (if any) may contain confidential, proprietary or legally privileged information and is intended only for the use of the addressee named above. No confidentiality or privilege is waived or lost by any mistransmission. If you are not the intended recipient of this message you are hereby notified that you must not use, disseminate, copy it in any form or take any action in reliance on it. If you have received this message in error please delete it and any copies of it and kindly inform the sender of this e-mail.

**Comment by DigitalFlow** [comment]
Hi @alexmare93,

thanks for trying to provide the pictures.
Unfortunately they seem to have been lost inside the Email response.
Can you upload them directly in the GitHub interface?

You can just drag and drop them to the response editor.

Kind Regards,
Florian

**Comment by alexmare93** [comment]
This one is from the default sitemap:
![CF1](https://user-images.githubusercontent.com/57076004/81905562-292afb00-95c5-11ea-9eea-6fff48541b17.jpg)

By going here
![CF2](https://user-images.githubusercontent.com/57076004/81905563-29c39180-95c5-11ea-850e-1c04bda252a7.jpg)

Then here:
![CF3](https://user-images.githubusercontent.com/57076004/81905564-29c39180-95c5-11ea-83a1-c1addce7bf02.jpg)

Something changes:
![CF4](https://user-images.githubusercontent.com/57076004/81905566-2a5c2800-95c5-11ea-82e5-3f7e78765961.jpg)

**Comment by andreaPorini** [comment]
Hello @DigitalFlow,

do you have any update on this issue? I work with @alexmare93 and we're still facing this issue.

Thank you for your support!

Best regards,
Andrea

**Comment by andreaPorini** [comment]
> Hello @DigitalFlow,
> 
> do you have any update on this issue? I work with @alexmare93 and we're still facing this issue.
> 
> Thank you for your support!
> 
> Best regards,
> Andrea

I think the problem in your JS `xrm_html_templating.js`, specifically here:
![image](https://user-images.githubusercontent.com/45461747/82565912-76860a00-9b7b-11ea-9b3f-055f42a762c5.png)

The next function `const config = search.indexOf("{") !== -1 ? JSON.parse(search.substring(6)) : undefined;` seems to break because the string "search" contains also the appid parameter, this is the content of the const: `"?appid=afb33431-aa91-418a-9a7d-6e1a1ca31100&data={"jsonField": "oss_json", "htmlField": "oss_html"}"`

If I try to execute the next step in the Chrome's Console, I get this error:
![image](https://user-images.githubusercontent.com/45461747/82566177-d11f6600-9b7b-11ea-85ee-d8c1d111b712.png)

Do you know if this could be the problem?

Best regards,
Andrea

**Comment by DigitalFlow** [comment]
I just merged #17 by @andreaPorini and released it as v1.4.1, thanks for the support.
Can you test whether this solves your issue?

Kind Regards,
Florian

**Comment by DigitalFlow** [comment]
Closing this, please reopen if issues persist for you.
